File: net.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import Queue
import collections
import SocketServer
import threading
import socket
import time
import resource

from imgio import read_frame, write_frame, read_jpeg_frame

QUEUE_MAXSIZE = 5 * 30


def clear_line():
    sys.stderr.write('\r\033[2K')


class ImageSocketServerHandler(SocketServer.BaseRequestHandler):

    def handle(self):
        # When beginning execution, register itself against the
        # ImageSocketServer
        self.queue = Queue.Queue(maxsize=QUEUE_MAXSIZE)
        self.thread = threading.current_thread()
        self.server.register_handler(self)
        self.name = "%s %s" % self.client_address

        print >> sys.stderr, "Start sending data to %s" % (self.name)

        # Then keep waiting for data and writing it to the socket
        try:
            while True:
                data = self.queue.get(block=True)
                self.request.sendall(data)
                # There apparently is no mechanism to flush the socket
                self.queue.task_done()

        except:
            print >> sys.stderr, "Exception while sending data to %s" % (self.name)

        print >> sys.stderr, "Finished sending data to %s" % (self.name)

    def finish(self):
        self.request.close()

    def post_data(self, data):
        # If the thread is dead, return False, so that the
        # ImageSocketServer stops sending data and performs cleanup
        if not self.thread.is_alive():
            return False

        # Else, enqueue the new data payload
        try:
            self.queue.put(data, block=False)
        except Queue.Full:
            print >> sys.stderr, "Frame lost when queueing for %s" % (self.name)

        return True


class ImageSocketServer(SocketServer.ThreadingTCPServer):

    def __init__(self, host, port):
        SocketServer.ThreadingTCPServer.__init__(self, (host, port), ImageSocketServerHandler)

        self.handlers = []
        self.daemon_threads = True
        self.main_thread = threading.Thread(target=self.serve_forever)
        self.main_thread.daemon = True
        self.main_thread.start()

    def register_handler(self, handler):
        self.handlers.append(handler)

    def post_data(self, data):
        # Pass the payload to all handlers
        to_remove = []
        for handler in self.handlers:
            alive = handler.post_data(data)
            if not alive:
                to_remove.append(handler)

        # Remove all handlers that have died (FIXME: this could be
        # more efficient)
        for handler in to_remove:
            self.handlers.remove(handler)


class ImageMultiClient:

    def __init__(self, servers):
        """Connect to a number of image servers.

        The parameter servers is a list of tuples of the type ((host,
        port), timeout). Timeout is a float value that specifies the
        maximum number of seconds frames getting in can delay behind
        real time.

        """
        # Lay out arguments in internal structures
        self.addresses = []
        self.timeouts = []
        for i, (address, timeout) in enumerate(servers):
            self.addresses.append(address)
            self.timeouts.append(timeout)

        # Individual frame queues and threads
        self.queues = [collections.deque() for _ in servers]
        self.threads = []

        # Every time a thread has a new frame, it notifies this
        # condition variable
        self.condition = threading.Condition()

        # Debugging and profiling
        self.memory_used = [None] * len(servers)

        # Set up and spawn threads
        for i, _ in enumerate(servers):
            thread = threading.Thread(target=self.worker, args=[i])
            thread.daemon = True
            self.threads.append(thread)
        for thread in self.threads:
            thread.start()

    def get_status(self):
        now = time.time()
        ret = []
        for queue, memory in zip(self.queues, self.memory_used):
            delay = None
            if len(queue) != 0:
                # [last item][timestamp]
                delay = now - queue[-1][1]
            ret.append((delay, len(queue), memory))
        return ret

    def write_status(self):
        status = self.get_status()
        clear_line()
        sys.stderr.write(("%40s " * len(status)) % tuple(status))

    def worker(self, shard):
        # Set up the connection
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect(self.addresses[shard])
        fin = sock.makefile()

        # Start receiving frames
        while True:
            image, timestamp = read_frame(fin)
            with self.condition:
                self.queues[shard].append((image, timestamp))
                self.condition.notifyAll()
            # RUSAGE_THREAD is not mapped in resource module, but
            # direct inspection of Linux headers shows it is 1
            self.memory_used[shard] = resource.getrusage(1).ru_maxrss / 1000

    def advance_to_timestamp(self, timestamp, empty=False):
        """Pop all elements with timestamp strictly smaller than the specified
        one, from all queues. As an exception, if empty is False then
        queues are never emptied; i.e., last item never gets deleted,
        even when it has past timestamp.

        """
        with self.condition:
            # Trim queues
            for queue in self.queues:
                while True:
                    if len(queue) == 1 and not empty:
                        break
                    try:
                        first = queue[0]
                    except IndexError:
                        break
                    # [timestamp]
                    if first[1] < timestamp:
                        queue.popleft()
                    else:
                        break

            # Build and return frames
            ret = []
            for queue in self.queues:
                if len(queue) == 0:
                    ret.append(None)
                else:
                    ret.append(queue[0])
            return ret

    def advance_to_stream(self, shard, block=False, empty=False):
        with self.condition:
            queue = self.queues[shard]

            # Discard first frame if there is one, so that we do not
            # go over the same frame over and over
            if len(queue) > 0:
                queue.popleft()

            # If blocking, wait to have another frame
            if block:
                while True:
                    if len(queue) > 0:
                        break
                    else:
                        self.condition.wait()

            # Check if there is a new frame (there is one for sure if
            # we were blocking)
            try:
                # [shard][first element][timestamp]
                timestamp = queue[0][1]
            except IndexError:
                return None

            # If there is, proceed with advancing
            return self.advance_to_timestamp(timestamp, empty=empty)


class WritingThread:
    """A simplified and non-socket-bound version of the classes
    above. Instead of writing data to a file, you pass them to a
    thread that writes to the file. The main program is not blocked by
    writing to the file.

    Just for testing.

    """

    def __init__(self, fout):
        self.fout = fout
        self.thread = threading.Thread(target=self.work)
        self.thread.daemon = True
        self.queue = Queue.Queue(maxsize=QUEUE_MAXSIZE)
        self.active = True

        self.thread.start()

    def work(self):
        try:
            while True:
                data = self.queue.get(block=True)
                self.fout.write(data)
                #self.fout.flush()
                self.queue.task_done()

        except:
            self.active = False
            raise

    def post_data(self, data):
        if not self.active:
            return False

        try:
            self.queue.put(data, block=False)
        except Queue.Full:
            print >> sys.stderr, "Frame lost when queueing"

        return True

    def write(self, data):
        self.post_data(data)

File: net.cpp

#include "net.hpp"

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <netdb.h>
#include <unistd.h>
#include <assert.h>

using namespace chrono;

ImageServerConf::ImageServerConf(string host, string port, double timeout)
  : host(host), port(port), timeout(timeout) {

}

ImageMultiClient::ImageMultiClient(const vector< ImageServerConf > &servers, ImageOp op)
  : servers(servers), op(op), running(true) {

  // Allocate queues and start threads; threads vector is initially
  // resized in order to be sure that no threads will access it while
  // it is being rehashed
  this->queues.resize(servers.size());
  this->mem_usage.resize(servers.size());
  this->threads.resize(servers.size());
  for (unsigned int i = 0; i < servers.size(); i++) {
    this->threads[i] = thread(&ImageMultiClient::worker, this, i);
  }

}

ImageMultiClient::~ImageMultiClient() {

  this->stop();
  for (auto &thread : this->threads) {
    thread.join();
  }

}

void ImageMultiClient::worker(int shard) {

  int sd = -1;
  TJContext *ctx = NULL;
  struct addrinfo *res = NULL;
  int succeed;
  struct addrinfo hints;
  FILE *fin;

  // Create TJContext
  ctx = create_tjcontext();
  if (ctx == NULL) {
    fprintf(stderr, "Worker %d, could not create TJContext", shard);
    goto cleanup;
  }

  // Resolve network name
  bzero(&hints, sizeof(struct addrinfo));
  hints.ai_family = AF_UNSPEC;
  hints.ai_socktype = SOCK_STREAM;
  succeed = getaddrinfo(this->servers[shard].host.c_str(), this->servers[shard].port.c_str(), &hints, &res);
  if (succeed != 0) {
    fprintf(stderr, "Worker %d, could not resolve network name: %s", shard, gai_strerror(succeed));
    goto cleanup;
  }

  // Create the socket
  sd = socket(AF_INET, SOCK_STREAM, 0);
  if (sd < 0) {
    fprintf(stderr, "Worker %d, could not create socket: %s", shard, strerror(errno));
    goto cleanup;
  }

  // Connect the socket
  succeed = connect(sd, res->ai_addr, res->ai_addrlen);
  if (succeed < 0) {
    fprintf(stderr, "Worker %d, could not connect socket: %s", shard, strerror(errno));
    goto cleanup;
  }

  // Create the C-style file
  fin = fdopen(sd, "r");
  if (fin == NULL) {
    fprintf(stderr, "Worker %d, could not open file from socket: %s", shard, strerror(errno));
    goto cleanup;
  }

  while (this->running) {
    // Take an image
    Image *image;
    if (this->op == IM_OP_DECODE) image = read_frame(ctx, fin);
    else if (this->op == IM_OP_RAW) image = read_jpeg_frame(ctx, fin);
    else if (this->op == IM_OP_YUV_DECODE) image = read_frame_to_yuv(ctx, fin);
    else assert(false);
    if (image == NULL) {
      fprintf(stderr, "Worker %d, could not read image (socket was probably closed)", shard);
      goto cleanup;
    }

    // Take the lock on queues and push the image to the right queue
    unique_lock< mutex > lock(this->queues_mutex);
    this->queues[shard].push_back(image);
    lock.unlock();

    // Update memory usage
    struct rusage resources;
    getrusage(RUSAGE_THREAD, &resources);
    this->mem_usage[shard] = resources.ru_maxrss;

    // Then notify the waiting threads
    this->queues_condition.notify_all();
  }

 cleanup:
  if (sd >= 0) {
    succeed = close(sd);
    if (succeed < 0) {
      fprintf(stderr, "Worker %d, could not close socket: %s (but I do not know what to do about that...)", shard, strerror(errno));
    }
  }
  free(res);
  free_tjcontext(ctx);

}

vector< const Image* > ImageMultiClient::advance_to_timestamp(double timestamp, bool block, bool empty) {

  // Acquire both locks (so that worker() is permitted to make
  // progress when advance_to_*() is waiting; but two calls to
  // advance_to_*() are never concurrently allowed)
  unique_lock< recursive_mutex > adv_lock(this->advance_mutex);
  unique_lock< mutex > lock(this->queues_mutex);

  // Trim queues and check if we have to wait for frames
  bool waiting = false;
  do {
    for (unsigned int i = 0; i < this->queues.size(); i++) {
      auto &queue = this->queues[i];

      // Trim the queue
      while (true) {
        if (queue.size() == 1 && !empty) break;
        if (queue.size() == 0) break;
        Image *first = *queue.begin();
        if (first->timestamp < timestamp) {
          free_image(first);
          queue.pop_front();
        } else {
          break;
        }
      }

      // Check whether the queue needs to be waiting for frames: we
      // need to wait if (after trimming) we have exactly one frame in
      // the queue and its timestamp is not too old (i.e., greater
      // than reference timestamp minus the queue timeout) and is not
      // greater than reference timestamp
      if (queue.size() == 1 && (*queue.begin())->timestamp >= timestamp - this->servers[i].timeout && (*queue.begin())->timestamp < timestamp) {
        waiting = true;
      }
    }

    // If we have to wait, let us block or return
    if (waiting) {
      if (!block) {
        return vector< const Image * >();
      } else {
        this->queues_condition.wait(lock);
      }
    }
  } while (waiting);

  // Build and return frames
  vector< const Image* > ret;
  for (auto it = this->queues.begin(); it != this->queues.end(); it++) {
    auto &queue = *it;
    if (queue.size() == 0) ret.push_back(NULL);
    else ret.push_back(*queue.begin());
  }

  return ret;

}

vector< const Image* > ImageMultiClient::advance_to_stream(int shard, bool block, bool empty) {

  // Acquire both locks (see advance_to_timestamp())
  unique_lock< recursive_mutex > adv_lock(this->advance_mutex);
  unique_lock< mutex > lock(this->queues_mutex);

  auto &primary = this->queues[shard];

  // If blocking, wait to have at least two frames in primary queue
  if (block) {
    while (primary.size() < 2) {
      this->queues_condition.wait(lock);
    }
  }

  // Check the timestamp of the second frame
  double timestamp;
  if (primary.size() >= 2) {
    timestamp = (*(++primary.begin()))->timestamp;
  } else {
    return vector< const Image * >();
  }

  // Unlock before entering advance_to_timestamp(), otherwise it would
  // deadlock
  lock.unlock();

  return advance_to_timestamp(timestamp, block, empty);

}

vector< const Image* > ImageMultiClient::advance_to_last(bool block, bool empty) {

  // Acquire both locks (see advance_to_timestamp())
  unique_lock< recursive_mutex > adv_lock(this->advance_mutex);
  unique_lock< mutex > lock(this->queues_mutex);

  double timestamp = 1e20;
  for (auto &queue : this->queues) {
    if (queue.size() > 0) {
      timestamp = min(timestamp, (*(--queue.end()))->timestamp);
    }
  }

  // Unlock before entering advance_to_timestamp(), otherwise it would
  // deadlock
  lock.unlock();

  return advance_to_timestamp(timestamp, block, empty);

}

vector< tuple< double, int, int > > ImageMultiClient::get_status() {

  unique_lock< mutex > lock(this->queues_mutex);
  double now = duration_cast< duration< double > >(system_clock::now().time_since_epoch()).count();

  vector< tuple< double, int, int > > ret;
  for (unsigned int i = 0; i < this->queues.size(); i++) {
    auto &queue = this->queues[i];
    double time_delta;
    if (queue.size() > 0) time_delta = now - (*(--queue.end()))->timestamp;
    else time_delta = -100.0;
    ret.push_back(make_tuple(time_delta, queue.size(), this->mem_usage[i]));
  }
  return ret;

}

void ImageMultiClient::print_status() {

  auto status = this->get_status();
  for (auto &s : status) {
    fprintf(stderr, "%f %d %d     ", get<0>(s), get<1>(s), get<2>(s));
  }
  fprintf(stderr, "\n");
  fflush(stderr);

}

void ImageMultiClient::stop() {

  this->running = false;

}
